slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,365 @@
+				obj, err := pm.kubeClient.Core().Events(v1.NamespaceAll).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = eventSelector
+				return pm.kubeClient.Core().Events(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Event{},
+		func(obj interface{}) error {
+			e, ok := obj.(*v1.Event)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Event: %v", obj)
+				return nil
+			}
+			pm.handleEventUpdate(e)
+			return nil
+		},
+		func(obj interface{}) error {
+			return nil
+		},
+	)
+	go eventController.Run(stopCh)
+	wait.Until(pm.purgeOutdated, 15*time.Second, stopCh)
+	return nil
+}
+
+func (pm *PodStartupLatencyDataMonitor) purgeOutdated() {
+	now := time.Now()
+	pm.Lock()
+	defer pm.Unlock()
+	if len(pm.toDelete) == 0 {
+		return
+	}
+	left := []string{}
+	for _, podKey := range pm.toDelete {
+		if data, ok := pm.PodStartupData[podKey]; ok {
+			if data.deletedTimestamp.Add(pm.purgeAfter).After(now) {
+				left = append(left, podKey)
+				continue
+			}
+			delete(pm.PodStartupData, podKey)
+		}
+	}
+
+	pm.toDelete = left
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodUpdate(p *v1.Pod) {
+	rv, err := storage.ParseListResourceVersion(p.ResourceVersion)
+	if err != nil {
+		glog.Errorf("Failed to parse resourceVersion %v from Pod %v. Skipping.", p.ResourceVersion, getPodKey(p))
+		return
+	}
+	if rv <= pm.startPodRV {
+		// Still processing backlog from before startup. We don't want to process this pod,
+		// as we depend on local clock to determine running timestamp, and doing this for pods
+		// that were already running before starting this component makes no sense.
+		pm.markAsAlreadyRunning(getPodKey(p))
+		return
+	}
+	if isReady, create, running := checkPodAndGetStartupLatency(p); isReady {
+		go pm.insertPodRunningTime(getPodKey(p), create, running)
+	}
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodDelete(p *v1.Pod) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	key := getPodKey(p)
+	data, ok := pm.PodStartupData[key]
+	if !ok {
+		data = PodStartupMilestones{}
+		data.startedPulling = time.Unix(0, 0)
+	}
+	data.deletedTimestamp = time.Now()
+	pm.PodStartupData[key] = data
+	pm.toDelete = append(pm.toDelete, key)
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePullingImageEvent(key string, e *v1.Event) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	ok := false
+	data := PodStartupMilestones{}
+	if data, ok = pm.PodStartupData[key]; !ok {
+		data.finishedPulling = time.Unix(0, 0)
+	}
+	data.startedPulling = e.FirstTimestamp.Time
+	pm.PodStartupData[key] = data
+
+	if !data.created.IsZero() && !data.startedPulling.IsZero() && !data.finishedPulling.IsZero() && !data.observedRunning.IsZero() {
+		glog.V(4).Infof("Observed Pod %v creation: created %v, pulling: %v, pulled: %v, running: %v",
+			key, data.created, data.startedPulling, data.finishedPulling, data.observedRunning)
+		pm.updateMetric(&data)
+	}
+}
```

Comment:
One thing I started wondering about - there seem to be a silent assumption here, that we are pulling one image for a pod. What if there are multiple containers in a single pod? We will do multiple pulls in such case. This doesn't seem to be handled here.